src/activatable/activatable.ts

```diff
@@ -0,0 +1,130 @@
+/**
+ * Created ->- Activating ->- Activated ->- Deactivating ->- Deactivated -------------->----------- Disposed
+ *                 |  |                                         |  |                                  |  |
+ *                 |  `------------------<----------------------'  `--->--- FailedToDeactivate --->---'  |
+ *                 `--------------------->--------- FailedToActivate ------------------>-----------------'
+ */
+
+enum LifeCycleState {
+  Created = 1,
+  Activating,
+  Activated,
+  Deactivating,
+  Deactivated,
+  Disposed,
+  FailedToActivate,
+  FailedToDeactivate,
+}
+
+const LifeCycleDescription = {
+  [LifeCycleState.Created]: 'has never activated before',
+  [LifeCycleState.Activating]: 'is activating',
+  [LifeCycleState.Activated]: 'is activated',
+  [LifeCycleState.Deactivating]: 'is deactivating',
+  [LifeCycleState.Deactivated]: 'is deactivated',
+  [LifeCycleState.Disposed]: 'is disposed',
+  [LifeCycleState.FailedToActivate]: 'errored while activating',
+  [LifeCycleState.FailedToDeactivate]: 'errored while deactivating',
+};
+
+export interface IActivatable {
+  isActivating: boolean;
+  isActivated: boolean;
+  isDeactivating: boolean;
+  isDeactivated: boolean;
+  isErrored: boolean;
+  activate: () => Promise<void>;
+  deactivate: () => Promise<void>;
+  disposeActivateable: () => void;
```

Comment:
```suggestion
  dispose: () => void;
```
How about a plain `dispose`? `IActivatable` could just extend `IDisposable` interface.

Reply:
I _really_ am like _really_ not sure if I even want this. I threw this in because technically it's needed to get to the final `Disposed` state but I'm not sure if this is the best way to do it.

One question I have is: should all `activatables` be `disposable`? In other words, _should_ `IActivatable` just extend the `IDisposable` interface?

Reply:
disposing is dead. no disposed state, no need to dispose an activatable.